src/OrbitGl/LineGraphTrack.cpp

```diff
@@ -52,30 +52,61 @@ void LineGraphTrack<Dimension>::DrawSeries(PrimitiveAssembler& primitive_assembl
-
-    DrawSingleSeriesEntry(primitive_assembler, current_time, next_time, current_normalized_values,
-                          next_normalized_values, z, is_last);
-
-    current_iterator = next_iterator;
-    current_time = next_time;
-    current_normalized_values = next_normalized_values;
+  std::array<float, Dimension> prev_normalized_values =
+      GetNormalizedValues(curr_iterator->second, min, inverse_value_range);
+
+  bool is_first = true;
+  GraphTrackDataAggregator<Dimension> aggr(this->aggregation_mode_);
+
+  const uint32_t resolution_in_pixels =
+      this->GetViewport()->WorldToScreen({this->GetWidth(), 0})[0];
+  uint64_t next_pixel_start_ns = orbit_client_data::GetNextPixelBoundaryTimeNs(
+      min_tick, resolution_in_pixels, min_tick, max_tick);
+
+  while (curr_iterator != last_iterator) {
+    uint64_t prev_time = curr_iterator->first;
+    curr_iterator = std::next(curr_iterator);
+
+    uint64_t curr_time = curr_iterator->first;
+    std::array<float, Dimension> curr_normalized_values =
+        GetNormalizedValues(curr_iterator->second, min, inverse_value_range);
+
+    if (is_first) {
+      aggr.StartNewEntry(prev_time, curr_time, curr_normalized_values);
+      is_first = false;
+    } else {
+      // If the size of the current entry is less than a pixel
+      if (aggr.GetEntry().end_tick < next_pixel_start_ns) {
+        // Add the current data to accumulated_entry
+        aggr.AppendData(prev_time, curr_time, curr_normalized_values);
+      } else {
+        // Otherwise, draw the accumulated_entry and start accumulating a new one
+        DrawSingleSeriesEntry(primitive_assembler, aggr.GetEntry().start_tick,
+                              aggr.GetEntry().end_tick, prev_normalized_values,
+                              aggr.GetEntry().values, z, false);
+
+        next_pixel_start_ns = orbit_client_data::GetNextPixelBoundaryTimeNs(
+            aggr.GetEntry().end_tick, resolution_in_pixels, min_tick, max_tick);
+
+        prev_normalized_values = aggr.GetEntry().values;
+        aggr.StartNewEntry(prev_time, curr_time, curr_normalized_values);
+      }
+    }
   }
 
-  if (current_time < max_tick) {
-    DrawSingleSeriesEntry(primitive_assembler, current_time, max_tick, current_normalized_values,
-                          current_normalized_values, z, true);
+  bool is_accumulated_entry_last = aggr.GetEntry().end_tick >= max_tick;
+  // Draw the leftover entry
+  DrawSingleSeriesEntry(primitive_assembler, aggr.GetEntry().start_tick, aggr.GetEntry().end_tick,
+                        prev_normalized_values, aggr.GetEntry().values, z,
+                        is_accumulated_entry_last);
+
+  // If there was not enough data to reach the end tick, draw an entry until the
+  // end.
+  if (!is_accumulated_entry_last) {
+    DrawSingleSeriesEntry(primitive_assembler, aggr.GetEntry().end_tick, max_tick,
+                          aggr.GetEntry().values, aggr.GetEntry().values, z, true);
   }
```

Comment:
If we draw in the previous way "current --> next" instead of "previous --> current", then we don't need the `is_first` and also we don't need to call `DrawSingleSeriesEntry` twice here

Reply:
The main problem is that if we're aggregating the entries then we don't know the next values until we've run the aggregation on them. And if we do aggregate them the algorithm becomes identical to this one